populartimes/crawler.py

```diff
@@ -106,28 +108,36 @@ def worker_radar():
 
 
 def get_radar(_lat, _lng):
-    # places - radar search - https://developers.google.com/places/web-service/search?hl=de#RadarSearchRequests
-    radar_str = radar_url.format(_lat, _lng, params["radius"], "|".join(params["type"]), params["API_key"])
-    resp = json.loads(requests.get(radar_str, auth=('user', 'pass')).text)
-    check_response_code(resp)
-    radar = resp["results"]
+    # places - nearby search - https://developers.google.com/places/web-service/search?hl=en#PlaceSearchRequests
+    radar_str = nearby_url.format(_lat, _lng, params["radius"], "|".join(params["type"]), params["API_key"])
+    curr_radar_str = radar_str
+    cnt_res = 0
+    while (True):
+        resp = json.loads(requests.get(curr_radar_str, auth=('user', 'pass')).text)
 
-    if len(radar) > 200:
-        logging.warning("more than 200 places in search radius, some data may get lost")
+        check_response_code(resp)
+        radar = resp["results"]
 
-    bounds = params["bounds"]
+        cnt_res += len(radar)
+        if cnt_res >= 60:
+            logging.warning("Result limit in search radius reached, some data may get lost")
 
-    # retrieve google ids for detail search
-    for place in radar:
+        bounds = params["bounds"]
 
-        geo = place["geometry"]["location"]
+        # retrieve google ids for detail search
+        for place in radar:
 
-        if bounds["lower"]["lat"] <= geo["lat"] <= bounds["upper"]["lat"] \
-                and bounds["lower"]["lng"] <= geo["lng"] <= bounds["upper"]["lng"]:
+            geo = place["geometry"]["location"]
+            if bounds["lower"]["lat"] <= geo["lat"] <= bounds["upper"]["lat"] \
+                    and bounds["lower"]["lng"] <= geo["lng"] <= bounds["upper"]["lng"]:
 
-            # this isn't thread safe, but we don't really care, since at worst, a set entry is simply overwritten
-            if place["place_id"] not in g_place_ids:
-                g_place_ids.add(place["place_id"])
+                # this isn't thread safe, but we don't really care, since at worst, a set entry is simply overwritten
+                g_place_ids[place["place_id"]] = place
+        if "next_page_token" in resp:
+            curr_radar_str = radar_str + "&pagetoken=" + resp["next_page_token"]
+            sleep(2) # There is a "short" delay between when a next_page_token is issued, and when it will become valid.
```

Comment:
This is horrible, we have to wait at least two seconds before the next page becomes available!! We should make the search radius very small to avoid this case (=> max 20 results)

Reply:
The smaller the search radius the higher the number of calls for each search. Possibly try an "if not" catch on the  request, leading to the sleep(2), that way it only gets called  when necessary

Reply:
Sleep is only called if there is a next page.
As long as there are many places in the search area it should not matter if we make next page requests or more single page requests. But in less populated areas this is indeed an issue. And implementing an adaptive approach with varying search radiuses seems quite complicated to me.
We could continue processing other areas and come back the the next page request later (e.g. put it to the end of the queue), but would need to take track of when the previous request was sent and possibly sleep for the remaining wait time. This assumes that the next page tokens stay valid while sending many different requests.